Drivers: DW1000: Extend IEEE802.15.4 config API to config channel parameters

jli157 opened the pull request:
A couple of features are added:

* Extended IEEE802.15.4 config API to have channel specific settings.
* Implement supporting long frame (1024 bytes)
* Implement enabling/disable smart TX power for 6.8Mbps
* Implement setting channel configs by using the new API.

jli157 committed:
drivers: dw1000: add long frame and smart power mode

Add configurations to enable/disable long frame mode
and TX smart power mode.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jukkar commented on drivers/ieee802154/ieee802154_dw1000_regs.h:
```diff
@@ -1873,10 +1873,10 @@
  *                  Channel: na  1  2  3  4  5 na  7
  */
-const uint8_t dwt_ch_to_cfg[] = {0, 0, 1, 2, 3, 4, 0, 5};
+static const uint8_t dwt_ch_to_cfg[] = {0, 0, 1, 2, 3, 4, 0, 5};
```
This is not good, now the array is placed in multiple .c files and occupying the space multiple times. Better to put the arrays in a .c file and just reference the values in there.

jli157 replied:
That is right, if they are referred multiple times, there will be multiple copies. I believe these constants are used only by `ieee802154_dw1000.c`. One option is to move all these constants to the `dw1000.c` file but it will look a bit messy; another option is to move them to another header file which will be included by `dw1000.c`.  Otherwise, they have to be declared in the `dw1000.c` when being used if they are put in a new `.c` file.

jukkar replied:
You could create a new .c file for these values, and then just add "extern var..." in this header file. Then no changes are needed in dw1000.c file.

jli157 replied:
Sounds good.

jli157 replied:
I feel it is still not a good way. Look at the following error:
```
zephyr/drivers/ieee802154/ieee802154_dw1000.c:1238:29: note: in expansion of macro 'ARRAY_SIZE'
 1238 |  if (rf_cfg->rx_shr_code >= ARRAY_SIZE(dwt_lde_repc_defs)) {
```
The array `dwt_lde_repc_defs[]` is defined in the header file without its size specified. So, the macro `ARRAY_SIZE` has no way to figure out its size unless the array size is specified in the header file. 

Another issue is that I have to keep the comments in two files. It seems reasonable to remove the comments in the C file but the context will lose. 

So, either adding the array size for all arrays whose size is not defined in the header file; or moving all constants to a new header file which will be only included by dw1000.c. Any suggestions?

jukkar commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1309,6 +1320,20 @@ static int dwt_configure_rf_phy(struct dwt_context *ctx)
+		sys_cfg |= DWT_SYS_CFG_PHR_MODE_11;
+	}
+
+
```
extra empty line can be removed

jukkar commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1584,6 +1548,15 @@ static int dw1000_init(struct device *dev)
 	LOG_INF("GPIO and SPI configured");
 
+	/* Copy the phy configuration table */
+	const struct dwt_phy_config *rf_cfg = dwt_phy_cfg_get(ctx->phy_cfg_id);
```
Please declare the variables in the beginning of the block. Also remove the empty line between variable setting and the if() statement.

jfischer-phytec-iot commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -1309,6 +1320,20 @@ static int dwt_configure_rf_phy(struct dwt_context *ctx)
+
+	if (rf_cfg->smart_power_en && rf_cfg->dr == DWT_BR_6M8) {
+		/* Enable smart TX power control for 6.8Mbps */
+		sys_cfg &= ~DWT_SYS_CFG_DIS_STXP;
```
This is default behavior according to the DW1000 Manual, "When DIS_STXP is 0 and the data rate is configured to 6.8 Mbps, Smart Tx power is enabled.".  This change seems to be redundant.

jfischer-phytec-iot commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -150,22 +129,7 @@ static const struct dwt_hi_cfg dw1000_0_config = {
 static struct dwt_context dwt_0_context = {
 	.dev_lock = Z_SEM_INITIALIZER(dwt_0_context.dev_lock, 1,  1),
 	.phy_sem = Z_SEM_INITIALIZER(dwt_0_context.phy_sem, 0,  1),
-	.rf_cfg = {
```
That is not the way this (and any 15.4) device should be configured. This is config per device and should remain so. PHY attributes like channel, data rate, PRF, SHR code would be set by 15.4 subsystem. No need for this change, it makes the driver unnecessary complicated. First 15.4 should be proper separated from the network subsystem and UWB support and appropriate services added there.

jli157 replied:
DW1000 does support multiple configurations and in the real use cases, the user does have the needs to change the channel configuration at runtime. I agree that adding the configuration table is the application level's scope and should not be put here. Is it reasonable to add a vendor specific 15.4 config type like `IEEE802154_CONFIG_CUSTOM` so that there is a way to change the channel configuration table at runtime?

jukkar replied:
> there is a way to change the channel configuration table at runtime?

There is a mgmt API for 802.15.4 that can be used to change configurations at runtime. See `include/net/ieee802154_mgmt.h` for details.

jli157 replied:
I've checked that. The API only supports changing channel, TX power,  address and PAN_ID. There is not an API supporting to configure other channel parameters, such as data rates, PRF and SHR code.

jukkar replied:
> The API only supports changing channel, TX power, address and PAN_ID. There is not an API supporting to configure other channel parameters, such as data rates, PRF and SHR code.

Yes, I know. I should have been more clearer, what I meant is that there is a generic API for changing the configuration, someone just needs to implement the options you mentioned in that API.

jli157 replied:
Can you guys review my current changes, still not good or better? 

@jukkar I think some parameters are specific to UWB, maybe they can be grouped together as a channel phy parameters? If you don't mind, I can add that mgmt APIs to ask you guys to review. I would like to have these functions so that I can use in my projects.

jfischer-phytec-iot replied:
I do not know why this driver get you attention, but I would really appreciate if you could stop adding any new features to it, especially if you are not familiar with this device. The new features does not help and are not necessary at this point of development. Just adding new things to the 15.4 subsystem does not help too. First the 15.4 should be made an independent subsystem, this would allow simpler, better and faster testing. Also one would usually use this device without 6lo. I have started to work on and used it to develop this driver, but it was not suitable for the upstream. I am still on it but I can not spend much time on it right now. Second step would be adding rudimental support for the UWB PHY (and it not just a channel number)...

jli157 replied:
Sorry for the bothering. I have used the device for a couple years but not related with 6lo. Basically we have followed the design DecaWave had done before, which could not be well aligned to Zephyr's architecture. Thanks for the explanation! And I would take this PR off from zephyr development tree, just keep it in my private git. I appreciate what you have done on this device and would like to leverage it as much as possible for the future projects. Let me know if I can be helpful for something.  Thank you!

jli157 commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -84,6 +84,9 @@ struct dwt_phy_config {
 	uint8_t tx_shr_code;	/* TX SHR preamble code */
 	uint32_t tx_shr_nsync;	/* PLEN index, e.g. DWT_PLEN_64 */
 
+	bool phr_mode_ext;	/* Extended PHR mode: long frame 1024 Byte */
```
I agree this in terms of complying with the 15.4 standard. However, I would argue that the device has the capability of transmitting long frame. Also, there is a setting `CONFIG_IEEE802154_RAW_MODE` for the user to choose not totally comply with 15.4 and directly process the data by the user. Having the choice doesn't mean the driver will not comply with 15.4.

jfischer-phytec-iot replied:
This device has capability of so much, but it does not make sense to add something just because it is possible. I do not know why CONFIG_IEEE802154_RAW_MODE is relevant here or not comply to anything.

jli157 replied:
In my previous project, we have used long frame and smart TX power.  Anyway, I can remove these two extensions to comply with IEEE802.15.4 standard. Thanks for the comments. 

How about the runtime configuration API for changing the channel configuration? Any comments?

jli157 committed:
drivers: dw1000: expose phy config struct

phy config struct can be used to define phy
configs in another place.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 committed:
api: ieee802154: define channel settings

Add channel settings whose contents are defined
by a specific radio chip.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 committed:
drivers: dw1000: api for changing channel config

Use `IEEE802154_CONFIG_CHANNEL_SETTINGS` to change the whole
channel configuration at one time.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 committed:
driver: dw1000: move all constants to a heder file

dw1000_regs.h will be used by applications. All
constants defined in the file shall be moved out
to avoid compiling issues.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jfischer-no reviewed (changes requested):
NACK to new features, see my comments
